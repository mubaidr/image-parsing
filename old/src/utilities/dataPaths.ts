import os from 'os'
import path from 'path'

const isDev: boolean = process.env.NODE_ENV === 'development'
const tmp: string = isDev ? path.resolve('.tmp') : os.tmpdir()
const home: string = path.join(os.homedir(), 'desktop')

const dataPaths = {
  design: path.resolve('__tests__', '_data', 'design.qr.svg'),
  designBarcode: path.resolve('__tests__', '_data', 'design.svg'),
  home,
  images: path.resolve('__tests__', '_data', 'images-qrcode'),
  imagesBarcode: path.resolve('__tests__', '_data', 'images-barcode'),
  key: path.resolve('__tests__', '_data', 'key.xlsx'),
  keyImage: path.resolve('__tests__', '_data', 'key.jpg'),
  questionsModel: isDev
    ? path.resolve('src/data/questions-model.json')
    : path.resolve('dist/data/questions-model.json'),
  result: path.resolve('__tests__', '_data', 'result.xlsx'),
  resultCompiled: path.resolve('__tests__', '_data', 'resultCompiled.xlsx'),
  root: path.resolve('.'),
  temp: tmp,
  tmp,
}

export { dataPaths }
