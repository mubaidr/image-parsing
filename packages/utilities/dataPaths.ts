import os from 'os';
import path from 'path';

const tmp = os.tmpdir();

export const DataPaths = {
  design: path.resolve('tests', '_data', 'design.qr.svg'),
  designBarcode: path.resolve('tests', '_data', 'design.svg'),
  home: os.homedir(),
  images: path.resolve('tests', '_data', 'images-qrcode'),
  imagesBarcode: path.resolve('tests', '_data', 'images-barcode'),
  key: path.resolve('tests', '_data', 'key.xlsx'),
  keyImage: path.resolve('tests', '_data', 'key.jpg'),
  result: path.resolve('tests', '_data', 'result.xlsx'),
  resultCompiled: path.resolve('tests', '_data', 'resultCompiled.xlsx'),
  root: path.resolve('.'),
  temp: tmp,
  tmp,
};
