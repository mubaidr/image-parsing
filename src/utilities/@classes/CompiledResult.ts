import uuid from 'uuid'

import { importExcelToJson } from '../excel'
import Result from './Result'

class CompiledResult {
  private id: string
  private lastSavedTime: Date | undefined

  private keys: Result[] = []
  private results: Result[] = []

  public constructor() {
    this.id = uuid()
  }

  public getKeyCount(): number {
    return this.keys.length
  }

  public getResultCount(): number {
    return this.results.length
  }

  public getKeys(): Result[] {
    return this.keys
  }

  public getResults(): Result[] {
    return this.results
  }

  public addKeys(key: Result | Result[]): CompiledResult {
    if (key instanceof Result) {
      this.keys.push(key)
    } else {
      this.keys.push(...key)
    }

    return this
  }

  public addResults(result: Result | Result[]): CompiledResult {
    if (result instanceof Result) {
      this.results.push(result)
    } else {
      this.results.push(...result)
    }

    return this
  }

  public compile(marks?: number, negativeMarks?: number): CompiledResult {
    if (this.keys.length === 0) throw 'Keys not loaded'
    if (this.results.length === 0) throw 'Results not loaded'

    for (let i = 0; i < this.keys.length; i += 1) {
      for (let j = 0; j < this.results.length; j += 1) {
        this.results[j].compile(this.keys[i], marks, negativeMarks)
      }
    }

    return this
  }

  public static loadFromExcel(src: string): CompiledResult {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const rows: any[] = importExcelToJson(src)
    const compiledResult = new CompiledResult()

    rows.forEach(row => {
      const result = Result.fromJson(row)

      if (result.isKey()) {
        compiledResult.addKeys(result)
      } else {
        compiledResult.addResults(result)
      }
    })

    return compiledResult
  }

  public addFromExcel(src: string): CompiledResult {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const rows: any[] = importExcelToJson(src)

    rows.forEach(row => {
      const result = Result.fromJson(row)

      if (result.isKey()) {
        this.addKeys(result)
      } else {
        this.addResults(result)
      }
    })

    return this
  }

  public static merge(compiledResults: CompiledResult[]): CompiledResult {
    const compiledResult = new CompiledResult()

    compiledResults.forEach(cr => {
      compiledResult.addKeys(cr.getKeys()).addResults(cr.getResults())
    })

    return compiledResult
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public export(): any[] {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const obj: any[] = []

    this.results.forEach(result => {
      obj.push(result.toJson())
    })

    return obj
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public static export(compiledResult: CompiledResult): any[] {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const obj: any[] = []

    compiledResult.results.forEach(result => {
      obj.push(result.toJson())
    })

    return obj
  }
}

export default CompiledResult
