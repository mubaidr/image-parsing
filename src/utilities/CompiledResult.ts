import { v4 as uuidv4 } from 'uuid'
import { importExcelToJson } from './excel'
import { Result } from './Result'

export class CompiledResult {
  id: string
  lastSavedTime: Date | undefined
  keys: Result[] = []
  results: Result[] = []

  constructor() {
    this.id = uuidv4()
  }

  static loadFromExcel(src: string): CompiledResult {
    const rows: unknown[] = importExcelToJson(src)
    const compiledResult = new CompiledResult()

    rows.forEach((row) => {
      const result = Result.fromJson(row)

      if (result.isKey()) {
        compiledResult.addKeys([result])
      } else {
        compiledResult.addResults([result])
      }
    })

    return compiledResult
  }

  addKeys(keys: Result[]): CompiledResult {
    this.keys.push(...keys)
    return this
  }

  addResults(results: Result[]): CompiledResult {
    this.results.push(...results)
    return this
  }

  getKeysAndResults(): Result[] {
    return [...this.keys, ...this.results]
  }

  compile(marks?: number, negativeMarks?: number): CompiledResult {
    if (this.keys.length === 0) throw new Error('Keys not loaded')
    if (this.results.length === 0) throw new Error('Results not loaded')

    for (let i = 0; i < this.keys.length; i += 1) {
      for (let j = 0; j < this.results.length; j += 1) {
        this.results[j].compile(this.keys[i], marks, negativeMarks)
      }
    }

    return this
  }

  sortResults(): CompiledResult {
    this.results.sort((a, b) => {
      return parseInt(a.rollNo || '0', 10) - parseInt(b.rollNo || '0', 10)
    })

    return this
  }

  static merge(compiledResults: CompiledResult[]): CompiledResult {
    const compiledResult = new CompiledResult()

    compiledResults.forEach((cr) => {
      compiledResult.addKeys(cr.keys).addResults(cr.results)
    })

    return compiledResult
  }
}
