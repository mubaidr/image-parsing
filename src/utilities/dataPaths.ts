import os from 'os'
import path from 'path'

const isDev: boolean = process.env.NODE_ENV === 'development'

const tmp: string = isDev ? path.resolve('.tmp') : os.tmpdir()
const home: string = path.join(os.homedir(), 'desktop')

const dataPaths = {
  design: path.resolve('_test_data', 'design.qr.svg'),
  designBarcode: path.resolve('_test_data', 'design.svg'),
  home,
  images: path.resolve('_test_data', 'images-qrcode'),
  imagesBarcode: path.resolve('_test_data', 'images-barcode'),
  key: path.resolve('_test_data', 'key.xlsx'),
  keyImage: path.resolve('_test_data', 'key.jpg'),
  questionsModel: path.resolve('src/data/questions-model.json'),
  result: path.resolve('_test_data', 'result.xlsx'),
  root: path.resolve('.'),
  temp: tmp,
  tmp,
}

export { dataPaths }
