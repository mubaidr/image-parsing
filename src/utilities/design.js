const cheerio = require('cheerio')
const fs = require('fs')

const OPTIONS = require('./question-options')

/**
 * Extracts position & dimensions of objects from SVG design File
 *
 * @param {String} designFilePath Path to the svg design file
 * @returns {Object} JSON object
 */
async function getDesignData(dir) {
  const designData = {
    questions: {},
  }
  const ROLL_NO_PATTERN = new RegExp(/rollnobarcode/gi)
  const QUESTION_PATTERN = new RegExp(/(q[1-9][0-9]?[ad])\b/gi)

  const $ = cheerio.load(fs.readFileSync(dir, 'utf8'))
  const svgViewBox = $('svg')[0].attribs.viewBox.split(' ')

  designData.width = Math.ceil(svgViewBox[2] - svgViewBox[0])
  designData.height = Math.ceil(svgViewBox[3] - svgViewBox[1])

  let gotRollNo = false
  let x
  let y
  let rx
  let ry
  let width
  let height
  let left
  let top

  const groups = $('g')
  for (let i = 0; i < groups.length; i += 1) {
    const group = groups[i]

    const title =
      $(group)
        .find('title')
        .first()
        .html()
        .trim()
        .toUpperCase() || ''

    if (!title) continue

    const isQuestionGroup = QUESTION_PATTERN.test(title)
    const isRollNoGroup =
      isQuestionGroup || gotRollNo ? false : ROLL_NO_PATTERN.test(title)

    if (!(isQuestionGroup || isRollNoGroup)) continue

    const transform = $(group)
      .attr('transform')
      .replace(/(translate)|\(|\)/gi, '')
      .split(',')
      .map(val => parseInt(val, 10) || 0)

    const rect = $(group)
      .find('rect')
      .first()

    left = parseInt(rect.attr('x'), 10) || 0
    top = parseInt(rect.attr('y'), 10) || 0

    rx = parseInt(rect.attr('rx'), 10) || 0
    ry = parseInt(rect.attr('ry'), 10) || 0

    x = left - rx + transform[0]
    y = top - ry + transform[1]

    width = parseInt(rect.attr('width'), 10) || 0 + rx
    height = parseInt(rect.attr('height'), 10) || 0 + ry

    if (isQuestionGroup) {
      const optionTitle = title.slice(-1)
      const questionNumber = title.slice(0, -1)

      if (!designData.questions[questionNumber]) {
        designData.questions[questionNumber] = {}
      }

      if (optionTitle === OPTIONS.A) {
        designData.questions[questionNumber].x1 = x - 1
        designData.questions[questionNumber].y1 = y - 1
      } else {
        // donot check last option to suport any number of options
        designData.questions[questionNumber].x2 = x + width + 1
        designData.questions[questionNumber].y2 = y + height + 1
      }
    } else if (isRollNoGroup) {
      designData.rollNo = { x1: x, y1: y, x2: x + width, y2: y + height }
      gotRollNo = true
    }
  }

  return designData
}

module.exports = {
  getDesignData,
}
