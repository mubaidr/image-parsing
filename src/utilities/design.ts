import { parse } from 'fast-xml-parser'
import { readFileSync } from 'fs'
import { DesignData, ItemInfo } from './@classes/WorkerManager'
import RegExpPatterns from './@enums/RegExpPatterns'
import { dataPaths } from './dataPaths'

function getDesignPathByID(id: string) {
  //TODO: implement internal store for designs with import/export functionality
  console.log(id)

  return dataPaths.designBarcode
}

export async function getDesignData(id: string): Promise<DesignData> {
  const filePath = getDesignPathByID(id)

  const { svg } = parse(readFileSync(filePath).toString(), {
    attributeNamePrefix: '',
    ignoreAttributes: false,
    parseNodeValue: true,
    parseAttributeValue: true,
    allowBooleanAttributes: true,
  })

  // get width, height from viewbox
  const [x1, y1, x2, y2] = svg.viewBox
    .split(' ')
    .map((i: string) => parseInt(i, 10))
  const svgWidth = x2 - x1
  const svgHeight = y2 - y1

  // prepare pattern matching reg expressions
  const PATTERN_OPTION = new RegExp(RegExpPatterns.OPTION, 'i')
  const PATTERN_BARCODE = new RegExp(RegExpPatterns.BARCODE, 'i')
  const PATTERN_QRCODE = new RegExp(RegExpPatterns.QRCODE, 'i')

  // for export
  let code: ItemInfo = { x: 0, y: 0, width: 0, height: 0 }
  const questions: {
    [key: string]: ItemInfo
  } = {}

  svg.g.forEach((group: any) => {
    const { title, transform, rect } = group

    if (!rect) return

    const { rx, ry } = rect
    let { x, y, width, height } = rect
    let xTransform = 0
    let yTransform = 0

    if (transform) {
      ;[xTransform, yTransform] = transform
        .replace(/(translate)|\(|\)/gi, '')
        .split(',')
        .map((val: string) => parseInt(val, 10) || 0)
    }

    x = Math.floor(x - rx + xTransform - 3)
    y = Math.floor(y - ry + yTransform - 3)
    width = Math.ceil(width + rx + 6)
    height = Math.ceil(height + ry + 6)

    if (PATTERN_OPTION.test(title)) {
      const questionNumber: string = title.slice(0, -1)

      if (!questions[questionNumber]) {
        questions[questionNumber] = {
          x: svgWidth,
          y: svgHeight,
          width: 0,
          height: 0,
        }
      }

      questions[questionNumber].width += width

      if (questions[questionNumber].x > x) {
        questions[questionNumber].x = x
      }

      if (questions[questionNumber].y > y) {
        questions[questionNumber].y = y
      }

      if (questions[questionNumber].width < width) {
        questions[questionNumber].width = width
      }

      if (questions[questionNumber].height < height) {
        questions[questionNumber].height = height
      }
    } else if (PATTERN_BARCODE.test(title) || PATTERN_QRCODE.test(title)) {
      code = { x, y, width, height }
    }
  })

  return {
    code,
    questions,
    width: svgWidth,
    height: svgHeight,
  }
}
