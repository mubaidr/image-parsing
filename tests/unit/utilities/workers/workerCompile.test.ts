import CompiledResult from '../../../src/utilities/@classes/CompiledResult'
import { dataPaths } from '../../../src/utilities/dataPaths'
import { getDesignData } from '../../../src/utilities/design'
import { readKey } from '../../../src/utilities/readKey'
import workerCompile from '../../../src/utilities/workers/workerCompile'

describe('workerCompile', () => {
  test('should compile just fine', async () => {
    const designData = getDesignData(dataPaths.designBarcode)
    const results = CompiledResult.loadFromExcel(dataPaths.result).getResults()
    const keys = await readKey(dataPaths.key)

    const compiledResult = workerCompile.start(
      {
        designData,
        results,
        keys,
      },
      false,
    )

    const result = compiledResult ? compiledResult.getResults()[0] : {}
    expect(result).toMatchSnapshot({
      id: expect.any(String),
      imageFile: expect.anything(),
    })
  })
})
