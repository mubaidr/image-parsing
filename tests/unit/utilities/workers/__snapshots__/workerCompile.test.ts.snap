// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workerCompile should compile just fine 1`] = `
Object {
  "answers": Object {
    "q1": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q10": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q11": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "q12": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q13": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "q14": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "q15": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "q16": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "q17": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "q18": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "q19": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "q2": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q20": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "q21": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q22": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q23": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "*",
    },
    "q24": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q25": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q26": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "q27": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "q28": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "q29": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "q3": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "q30": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "q31": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "q32": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "q33": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q34": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "q35": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "q36": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "q37": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "q38": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q39": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "q4": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "q40": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "q41": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "q42": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "q43": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "q44": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "q45": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "q46": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q47": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "q48": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "q49": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "q5": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "q50": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q51": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q52": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "q53": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "q54": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q55": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q56": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q57": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "q58": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "B",
    },
    "q59": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q6": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q60": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "D",
    },
    "q7": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
    "q8": Object {
      "correct": true,
      "skipped": false,
      "unattempted": false,
      "value": "A",
    },
    "q9": Object {
      "correct": false,
      "skipped": false,
      "unattempted": false,
      "value": "C",
    },
  },
  "correctCount": 21,
  "id": Any<String>,
  "imageFile": Anything,
  "incorrectCount": 39,
  "isCompiled": true,
  "isRollNoExtracted": false,
  "marks": 0,
  "post": 1,
  "questionPaperType": 1,
  "rollNo": undefined,
  "skippedCount": 0,
  "testCenter": 1,
  "testTime": 1,
  "totalMarks": 0,
  "unattemptedCount": 0,
}
`;
